File: vox/swift/vox-runtime/Sources/VoxRuntime/SessionResume.swift
import Foundation

private let sessionResumeMetadataFlagsNone: UInt64 = 0

public let sessionResumeKeyMetadataKey = "vox-session-key"

public func appendSessionResumeKeyMetadata(
    _ metadata: [MetadataEntry],
    key: [UInt8]
) -> [MetadataEntry] {
    metadata + [
        MetadataEntry(
            key: sessionResumeKeyMetadataKey,
            value: .bytes(key),
            flags: sessionResumeMetadataFlagsNone
        )
    ]
}

public func metadataSessionResumeKey(_ metadata: [MetadataEntry]) -> [UInt8]? {
    for entry in metadata where entry.key == sessionResumeKeyMetadataKey {
        if case .bytes(let key) = entry.value, key.count == 16 {
            return key
        }
    }
    return nil
}

func freshSessionResumeKey() -> [UInt8] {
    var generator = SystemRandomNumberGenerator()
    return (0..<16).map { _ in UInt8.random(in: UInt8.min...UInt8.max, using: &generator) }
}

func sessionResumeKeysEqual(_ lhs: [UInt8], _ rhs: [UInt8]) -> Bool {
    lhs == rhs
}

File: vox/typescript/packages/vox-core/src/session_resume.ts
import type { Metadata } from "@bearcove/vox-wire";

export const SESSION_RESUME_KEY_METADATA_KEY = "vox-session-key";

export function appendSessionResumeKeyMetadata(metadata: Metadata, key: Uint8Array): Metadata {
  return [
    ...metadata,
    {
      key: SESSION_RESUME_KEY_METADATA_KEY,
      value: { tag: "Bytes", value: key.slice() },
      flags: 0n,
    },
  ];
}

export function metadataSessionResumeKey(
  metadata: Metadata,
): Uint8Array | null {
  for (const entry of metadata) {
    if (entry.key !== SESSION_RESUME_KEY_METADATA_KEY) {
      continue;
    }
    if (entry.value.tag === "Bytes" && entry.value.value.length === 16) {
      return entry.value.value.slice();
    }
  }
  return null;
}
